posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+        userAlgebra.updateAge(username, userUpdate.age) *> Ok(username)
+      }
+  }
+
+}
+```
+
+Now this particular implementation is missing a very important part: error handling. If we use the `UserAlgebra`'s interpreter previously defined we will clearly miss the three errors defined by the `UserError` ADT.
+
+***NOTE: If you are not familiar with these concepts make sure you check out [my talk at Scala Matsuri](https://youtu.be/pGfj_l-h3M8?t=887) early this year where I also talk about error handling in http applications using the Http4s library.***
+
+### Http Error Handling
+
+Okay let's just go ahead and add some error handling to our http route by taking advantange of the `MonadError` instance defined by our constraint `Sync[F]` and making use of the syntax provided by `cats`:
+
+```tut:book:silent
+class UserRoutesAlt[F[_]: Sync](userAlgebra: UserAlgebra[F]) extends Http4sDsl[F] {
+
+  import JsonCodecs._
+
+  val routes: HttpRoutes[F] = HttpRoutes.of[F] {
+
+    case GET -> Root / "users" / username =>
+      userAlgebra.find(username).flatMap {
+        case Some(user) => Ok(user)
+        case None => NotFound(username)
+      }
+
+    case req @ POST -> Root / "users" =>
+      req.as[User].flatMap { user =>
+        userAlgebra.save(user) *> Created(user.username)
+      }.handleErrorWith { // compiles without giving you "match non-exhaustive" error
+        case UserAlreadyExists(username) => Conflict(username)
+      }
+
+    case req @ PUT -> Root / "users" / username =>
+      req.as[UserUpdateAge].flatMap { userUpdate =>
+        userAlgebra.updateAge(username, userUpdate.age) *> Ok(username)
+      }.handleErrorWith { // compiles without giving you "match non-exhaustive" error
+        case InvalidUserAge(age) => BadRequest(s"Invalid age $age")
+      }
+  }
+
+}
+```
+
+Now we can say this implementation is quite elegant! We are handling and mapping business errors to the according http response and our code compiles without any warning whatsoever. But wait... We are not handling the `UserNotFound` error and the compiler didn't tell us about it! That's not cool and we as functional programmers believe in types because we can know what a function might do just by looking at the types but here it seems we hit the wall.
+
+The problem is that our constraint of type `Sync` from `cats-effect` has a `MonadError` instance with its type error fixed as `Throwable`. So the compiler can't help us here since this type is too generic. And we can't add a constraint for `MonadError[F, UserError]` because we would get an "ambigous implicits" error with two instances of `MonadError` in scope.
+
+So, what can we do about it?
+
+### Next level MTL: Optics
+
+I heard sometime ago about `Classy Optics` (Lenses, Prisms, etc) when I was learning `Haskell` and watched [this amazing talk](https://www.youtube.com/watch?v=GZPup5Iuaqw) by George Wilson but I never got to use this concept in Scala until now!
```

Comment:
Haskell isn't a code literal, it's just a proper noun.  Classy Optics is just a noun, too.